docs/design/2018-07-22-enhance-propagations.md

```diff
@@ -0,0 +1,189 @@
+
+For now, most of the constraint propagation work in TiDB is done by `propagateConstantSolver`, it does:
+
+1. Find `column = constant` expression and substitue the constant for column, as well as try to fold the substituted constant expression if possible, for example:
+
+ Given `a = b and a = 2 and b = 3`, it becomes `2 = 3` after substitution and lead to a final `false` constant.
+
+2. Find `column A = column B` expression(which happens in `join` statements mostly) and propagate expressions like `column op constant`(as well as `constant op column`) based on the equliaty relation, the supported operators are:
+
+ * `ast.LT('<')`
+ * `ast.GT('>')`
+ * `ast.LE('<=')`
+ * `ast.GE('>=')`
+ * `ast.NE('!=')`
+
+The `propagateConstantSolver` makes more detailed/explicit filters/constraints, which can be used within other optimization rule. For example, in predicate-pushdown optimization, it generates more predicates that can be pushed closer to the data source(TiKV), and thus reduce the amount of data in the whole data path.
+
+We can further do the optimization by introduce more rules and infer/propagate more constraints from the existings ones, which helps us building better logical plan.
+
+## Proposal
+
+Here are proposed rules we can consider:
+
+1. Infer more filters/constraints from column equality relation
+
+ We should be able to infer set of data constraints based on column equality, that is, any constraint on `a` now applies to `b` as long as:
+
+ * The constraint is deterministic
+ * The constraint doesn’t have any side effect
+
+ For example, 
+
+ `t1.a = t2.a and t1.a < 5` => `t2.a < 5`
+
+ `t1.a = t2.a and t1.a in (12, 13) and t2.a in (14, 15)` => `t1.a in (14, 15) and t2.a in (12, 13)`
+
+ `t1.a = t2.a and cast(t1.a, varchar(20) rlike 'abc'` => `cast(t1.a, varchar(20)) rlike 'abc'`
+  
+ Equality propagation should also be included:
+
+ `t1.a = t2.a and cast(t1.a, varchar(20) = 5` and` => `cast(t2.a, varchar(20)) = 5`
+
+ But following predicates cannot be propagated:
+
+ `t1.a = t2.a and t1.a < random()` -- the expression is non-deterministic
+
+ `t1.a = t2.a and t1.a < sleep()` -- the expression has side effect
+
+2. Infer NotNULL filters from comparison operator
+
+ We can infer `NotNULL` constraints from a comparison operator that doesn’t accept NULL, then we can know that related columns cannot be NULL and add `NotNull` filter to them:
+
+ `t1.a = t2.a` => `not(isnull(a)) and not(isnull(b))`
+
+ `t1.a != t2.a` => `not(isnull(a)) and not(isnull(b))`
+
+ `a < 5` => `not(isnull(a))`
+
+ NOTE: Those columns should not have `NotNULL` constraint attribute, or the inferred filters are unnecessary.
+
+3. Fold the constraints based their semantics to avoid redundancy 
+
+ After the propagations we may produce some duplicates predicates, and they can be combined. We should analyze all of the conditions and try to make the predicates clean:
+
+ `a = b and b = a` -> `a = b`
+
+ `a < 3 and 3 > a` -> `a < 3`
+
+ `a < 5 and a > 5` -> `False`
```

Comment:
Do you mean constructing an expression which is always `flase`, for example, `a = null`?

Reply:
I think replacing them with a constant `False` is fine